### DISCUSSION TITLE
Dictionary-based transformation for same transforms on images and labels

### DISCUSSION THREAD

**Discussion by elahe-badali** [Project-MONAI/MONAI#2827] [Q&A] [answered]
Hello @Nic-Ma, I hope you well.
I want to use dictionary based transformation to get same transform on both images and labels. My dataset images are 3d cardiac MRI images and labels are binary masks.
But it seems that is not working. 
My code :
```
train_transforms = Compose(
    [   AddChanneld(keys=["image", "label"]),
        RandFlipd(keys=["image", "label"]),
        RandRotate90d(keys=["image", "label"],prob=0.10 ),
        ToTensord(keys=["image", "label"]),
    ]
)

train_dataset = Dataset( data=train_dataset, transform=train_transforms )

sample_num = 110
img = train_dataset[sample_num]['image'].squeeze(0)
lbl = train_dataset[sample_num]['label'].squeeze(0)
fig , (ax1 , ax2 , ax3) = plt.subplots(1 , 3 , figsize = (8 , 8))
ax1.imshow(img[1 , : , :])
ax2.imshow(img[2 , : , :])
ax3.imshow(lbl)
```
When i run it, i get this results ( first and second slice of MRI and label showed):
![image](https://user-images.githubusercontent.com/39738835/130571832-1f2d04df-4756-4caa-ba5a-fc18945a7289.png)

and when run show code again i get this:
![image](https://user-images.githubusercontent.com/39738835/130572019-98e291ea-df1c-4367-a390-e3732dc17766.png)

I have 2 questions:
First, It seems that "Flip" is worked just on labels. Why the same transformation is not applied two both images and labels?
Second, Why there are two different results for the same code?

**Comment by rijobro** [comment, marked as answer]
Hi, every time you access an element in the dataset, the transforms will be randomised. Hence, in this code snippet:

```python
img = train_dataset[sample_num]['image'].squeeze(0)
lbl = train_dataset[sample_num]['label'].squeeze(0)
```
there is a different random state between the `img` and `lbl`, and therefore different transforms are applied to them. If you instead do:

```python
data = train_dataset[sample_num] 
img = data['image'].squeeze(0)
lbl = data['label'].squeeze(0)
```
this will hopefully sort your problem.

**Reply by rijobro** [reply]
If you want to get the same results every time you execute the code, you need to set the determinism. This can be done with:

```python
SEED = 0  # can be anything
monai.utils.set_determinism(SEED)
```
or
```python
train_transforms.set_random_seed(SEED)
```

**Reply by elahe-badali** [reply]
Hi, thank you for your comments. I tested both of your solutions but problem still persists.

**Reply by rijobro** [reply]
I'm attaching a screenshot of my first point. The first row of images call the dataset twice, and so get two different random states. The second set of images call the dataset just once, and have the same transform applied to both of them.

If it's still not working for you, could you please create a minimal working example. It should be self-contained, so use random image generation if necessary.

![Screenshot from 2021-08-24 10-18-39](https://user-images.githubusercontent.com/33289025/130591450-3ab10c05-14a8-4989-a880-bef8d2401491.png)

**Reply by elahe-badali** [reply]
As a toy example i coded similar to yours and it worked. But if i stack 25 of image (like MRI) the code didn't work:
 
![image](https://user-images.githubusercontent.com/39738835/130603608-466bc347-9339-443e-b0b9-b52667dfe77d.png)

**Reply by rijobro** [reply]
In the screenshot you have `lbl = img`, which I imagine should be `lbl = images`?

**Reply by elahe-badali** [reply]
In fact in my main dataset "label" is a 2D binary image and "image" is 3D MRI image. In my screenshot, just for the sake that "img" is 2D image i used it as label too(just for example) and stacked of "img"s makes 3D MRI image (again for example).

**Reply by rijobro** [reply]
Your label has no channel dimension. What happens if you do `lbl = img[None]`?

**Reply by elahe-badali** [reply]
I added this line. but still something irregular is happening. 

In some runs i get this:

![image](https://user-images.githubusercontent.com/39738835/130731274-9f6cf8c9-b91f-4e77-bfac-d7da2e82cdad.png)



But in most other cases every thing is ok!

![image](https://user-images.githubusercontent.com/39738835/130731199-945a1959-85e9-462c-a659-7f7af043715a.png)

I just want to make sure that the transformation that is done on the images is also applied to the labels.
What is the best way to do it?

**Reply by rijobro** [reply]
I think you have a few mistakes in your code that are adding confusion to the matter. I've made you a jupyter notebook that I hope illustrates the point that we're testing. `lbl` and `img` are 2d and 3d, respectively. They have the respective sizes of (1, 100, 100) and (25, 100, 100). I've used the transform `RandRotated`.

In the first set of imshows, the dataset is called twice, and so the transforms have different rotations relative to each other. In the second, the dataset was called just once, and no matter how many times you execute the bottom cell, the resulting `img` and `lbl` should look the same. Hopefully this clears some confusion for you.

The file is attached as a .txt, but it is actually a jupyter notebook (.ipynb). Just change the extension and you should be able to run it.

[dict_transforms.txt](https://github.com/Project-MONAI/MONAI/files/7046501/dict_transforms.txt)

**Reply by rijobro** [reply]
Here is a screenshot from my attached jupyter notebook
![Screenshot from 2021-08-25 12-19-52](https://user-images.githubusercontent.com/33289025/130781675-89d0b042-6ec8-43fb-97d6-3174ff8b83dc.png)

**Reply by elahe-badali** [reply]
Thank you for sharing jupyter file. I run it and it worked. I think i find where the problem comes from. I use "Compose" to create transform.
Could i ask to you add below lines to your jupyter and run it?

This lines as transform:
```
from monai.transforms import RandRotated , Compose, AddChanneld, RandRotate90d, ToTensord

transform = Compose(
    [
        AddChanneld(keys=["img" , "lbl"]),
        RandRotate90d(keys=["img", "lbl"], prob=0.7),
        ToTensord(keys=["img", "lbl"]),
    ]

)

```

And  this line to show (i just add "squeeze" to remove added channel):

```
d = ds[0]
img = d["img"].squeeze(0)
lbl = d["lbl"].squeeze(0)
print(img.shape, lbl.shape)
imshow(img[0], lbl[0])
```

And run last cell multiple times to see results. In some case i get :

![image](https://user-images.githubusercontent.com/39738835/130787365-49c84614-6f62-4655-9cac-085c4986f48b.png)

If my transformation definition has problems, Could you please code correct one with "Compose"?

**Reply by rijobro** [reply]
Hi, I don't know how you intend to add a channel to your image given that it already has the channel dimension attached. If the image is (25, 100, 100), where number of channels = 25, then you can't use that in conjunction with `AddChanneld`.

I took my previously attached notebook and wrapped the `RandRotated` in a Compose operator:
```python
Compose([
    RandRotated(["img", "lbl"], prob=1.0, range_x=np.pi, keep_size=True)
])
```
and the results were the same.

I'm not sure what you're doing in terms of squeezing and indexing -- `lbl.squeeze(0)` will remove the channel dimension, as will `lbl[0]`, so I think there might be some errors in your additions to the notebook. Try and start again with my example and add things in piece-by-piece until you can convince yourself that if you access an element of the dataset in one go, the label and image will have the same random instance of the transforms applied to them.

**Reply by elahe-badali** [reply]
For example take a look at https://pytorch.org/docs/stable/generated/torch.nn.Conv3d.html. Conv3d's input has the shape (N,C,D,H,W). In my case 25 is not showing channels. 25 is "D" in the input shape and shows depth of 3d image. "C" is channel and in my dataset because of 25 gray images i want set the channel to 1. For this, i used "AddChanneld" to set input shape to (1,25,64,64) and later in loader added batch dimension to it.

**Reply by rijobro** [reply]
Ok, I think you're saying that your image is like this:
    `C=1, D=25, H=64, W=64`
and your label is like this:
    `C=1, H=64, W=64`.
The problem is that if you apply a rotation around the first spatial dimension, and in your image the first spatial dimension, (D=25) is different from the first spatial dimension of your label (H=64), then a rotation around the first spatial axis will give you different results.

Why don't you try moving the depth component to the end of the image. In this way, you'll have image:
    `C=1, H=64, W=64, D=25`
and label:
    `C=1, H=64, W=64, D=1`

**Reply by elahe-badali** [reply]
I think it's too late but thank you @rijobro, your last comment fix the problem.
